File: CrackingTheCodeInterview/Python/Chapter1/p5_string_compress.py
__author__ = 'valenc3x'

def string_compress(word=None):
    compressed = str()
    current = str()
    for i, letter in enumerate(word):
        if current != letter:
            count = 0
            current = letter
        count += 1
        if i < len(word)-1 and letter != word[i+1] or i == len(word) -1:
            compressed = '%s%s%d' % (compressed, letter, count)

    return compressed if len(compressed) < len(word) else word


if __name__ == '__main__':
    print(string_compress('aaabcccdeee'))
    print(string_compress('ab'))
    print(string_compress('abcdeeeeeee'))
    print(string_compress('aaabbbc'))

File: CrackingTheCodeInterview/C#/Chapter1/P5 StringCompress.cs
using System;

namespace Chapter1
{
    public class CompressWords{

        public static string StringCompress(string word){
            string compressed = "";
            int count = 0;
            char current = '\0';
            for (int i = 0; i < word.Length; i++)
            {
                if (current != word[i])
                {
                    count = 0;
                    current = word[i];
                }
                count++;
                if (i <= word.Length-1 && current != word[i+1])
                {
                    compressed += string.Concat(current, count);
                }
            }
            if (compressed.Length > word.Length)
            {
                return word;
            }else{
                return compressed;
            }
        }

        public static void Main(){
            System.Console.WriteLine(StringCompress("a"));
            System.Console.WriteLine(StringCompress("abbbb"));
            System.Console.WriteLine(StringCompress("aaabbbcccaab"));
            System.Console.WriteLine(StringCompress("abcde"));
        }

    }
}
